### DISCUSSION TITLE
How to recover from a "install retries exhausted"

### DISCUSSION THREAD

**Discussion by cer** [fluxcd/flux2#3851] [Q&A]
One of the consequences of [missing HelmRepository secret](https://github.com/fluxcd/flux2/discussions/3850) is because those `HelmReleases` were not installed, a `HelmRelease` from a different `HelmRepository` was not able to become ready.

As a result, that HelmRelease ended up in the state "install retries exhausted".

To fix the problem I had to use `flux suspend/resume`. But why? Is this documented somewhere?

**Comment by stefanprodan** [comment]
Bumping the retries would fix it `spec.install.remediation.retries`. You could also set the retries to `-1` and this will make Flux retry forever.

Docs: https://fluxcd.io/flux/components/helm/helmreleases/#configuring-failure-remediation

**Reply by cer** [reply]
Thanks for the tip. 

But let's imagine that it "install retries exhausted" occurs.
Is `flux suspend/resume` the correct way to tell FluxCD to try again?
And, is this (state machine model + how to recover from failure) documented somewhere?

**Reply by stefanprodan** [reply]
Any changes to `.spec` of a HelmRelease resets the retry count, toggling `.spec.suspend` or changing something in the `.spec.values` has the same effect. This behavior is alluded to in the docs, maybe we need to call it out under failure remediation section cc @hiddeco

**Reply by cer** [reply]
In my scenario, HelmRelease X wasn't installed  successfully - pod couldn't become ready - because it was depended upon HelmReleases Y and Z being installed successfully.

As a result, the solution didn't involve changing X. Instead, I had to tell Flux to try again with X.

**Reply by hiddeco** [reply]
In a future release there will be specific annotations to be able to control the reset of the failure count, or a one off "forced" retry.

This is however pending some bigger architectural changes, for which I am trying (very hard) to accumulate time for at the moment.

**Comment by youcefguichi** [comment]
hey @stefanprodan, i was reproducing the `install retries exhausted`  and `upgrade retries exhausted` problem.

however when i fix the helmrelease values on my gitrepo, flux sync automatically and it doesn't stuck on the `install retries exhausted` state, i'm just curious if this would be always the case ? or it's not consistent and it could stuck sometimes ?

Or it used to stuck but you guys improved the helm controller lately ?

**Reply by hiddeco** [reply]
It gets stuck when it's out of retries according to the defined `.spec.*.remediation.retries` setting. Changing values will reset the counter, which may result in it then succeeding in e.g. one try, as described by Stefan in https://github.com/fluxcd/flux2/discussions/3851#discussioncomment-5840482.

**Reply by youcefguichi** [reply]
Thanks @hiddeco for your reply, it helps us a lot on the investigation that we were doing about flux's helm controller and how it works and deals with errors, we tried to put  together in a simple article the lessons we gained from our experience of the `upgrade retries exhausted` issue.

the article: https://gimlet.io/blog/error-handling-in-helm-controller-how-to-solve-the-infamous-upgrade-retries-exhausted-error

**Reply by smeeklai** [reply]
Great article

**Reply by pavel-mikhalchuk** [reply]
Hey @youcefguichi ,
The link is unfortunately broken. I'll add the archive version here - https://web.archive.org/web/20250213232116/https://gimlet.io/blog/error-handling-in-helm-controller-how-to-solve-the-infamous-upgrade-retries-exhausted-error
